raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -375,21 +418,48 @@ private BeginQuorumEpochResponseData buildBeginQuorumEpochResponse(Errors error)
                 .setLeaderId(quorum.leaderIdOrNil());
     }
 
-    private BeginQuorumEpochResponseData handleBeginQuorumEpochRequest(BeginQuorumEpochRequestData request) throws IOException {
-        Optional<Exception> errorOpt = validateLeaderOnlyRequest(request.leaderId(), request.leaderEpoch());
+    /**
+     * Handle a BeginEpoch request. This API may return the following errors:
+     *
+     * - {@link Errors#BROKER_NOT_AVAILABLE} if this node is currently shutting down
+     * - {@link Errors#INCONSISTENT_VOTER_SET} if the request suggests inconsistent voter membership (e.g.
+     *      if this node or the sender is not one of the current known voters)
+     * - {@link Errors#FENCED_LEADER_EPOCH} if the epoch is smaller than this node's epoch
+     */
+    private BeginQuorumEpochResponseData handleBeginQuorumEpochRequest(
+        RaftRequest.Inbound requestMetadata
+    ) throws IOException {
+        BeginQuorumEpochRequestData request = (BeginQuorumEpochRequestData) requestMetadata.data;
+        Optional<Errors> errorOpt = validateVoterOnlyRequest(request.leaderId(), request.leaderEpoch());
```

Comment:
I actually agrees with your previous TODO marker inside this function to not call `becomeUnattachedFollower` inside, and one reason I had is that here, we will call `becomeUnattachedFollower` first and then `becomeFetchingFollower` immediately after.

What about this (admittedly it is also a bit awkward): let this function returns `UNKNOWN_LEADER_EPOCH` if `requestEpoch > quorum.epoch()`, and the caller can decide whether to really encode this error or not --- e.g. in this caller we would just ignore it and call `becomeFetchingFollower` afterwards, whereas in the `handleVoteRequest` caller above we would then call `becomeUnattachedFollower` in the caller.

Reply:
Let me see if some name changes can make the `becomeUnattachedFollower` transition less surprising. Note that semantically there is no problem with a call to `becomeUnattachedFollower` followed by `becomeFetchingFollower`.